### PR TITLE
xdg-ninja 0.2.0.1 (new formula)

### PR DESCRIPTION
**Pull request by superatomic**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

[`xdg-ninja` is a tools that checks your $HOME for unwanted files and directories](https://github.com/b3nj5m1n/xdg-ninja)

### CODE REVIEW

File: Formula/xdg-ninja.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+  depends_on "jq"
+
+  def install
+    prefix.install "programs/"
+    prefix.install "xdg-ninja.sh" => "xdg-ninja"
+    bin.install_symlink prefix/"xdg-ninja"
+  end
+
+  test do
+    system bin/"xdg-ninja"
```
Comment (carlocab) [inline Formula/xdg-ninja.rb:19]: Can we check output here?

File: Formula/xdg-ninja.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  url "https://github.com/b3nj5m1n/xdg-ninja/archive/refs/tags/v0.2.0.1.tar.gz"
+  sha256 "f4f9ab4500e7cf865ff8b68c343537e27b9ff1e6068cb1387e516e608f77cec8"
+  license "MIT"
+  head "https://github.com/b3nj5m1n/xdg-ninja.git", branch: "main"
+
+  depends_on "glow"
+  depends_on "jq"
+
+  def install
+    prefix.install "programs/"
```
Comment (carlocab) [inline Formula/xdg-ninja.rb:13]: I wonder if it makes more sense to install this into `pkgshare`.

**Review by carlocab — COMMENTED** [review]
Why not also install `xdgnj`?

**Comment by superatomic** [comment]
> Why not also install `xdgnj`?

`xdgnj` is a developer tool for contributing to [`xdg-ninja`](https://github.com/b3nj5m1n/xdg-ninja). As such, the average user of `xdg-ninja` will never use `xdgnj`.

I would assume that most contributors would prefer to clone the `xdg-ninja` repository to contribute instead of installing `xdgnj` via Homebrew as a standalone without the repository (the `xdgnj` tool might not even be useful for contributing if it is installed with Homebrew).

If you believe that this formula should provide the development tools for `xdg-ninja` alongside of `xdg-ninja` itself, I can try to add it regardless, even though I believe it probably shouldn't be installed.

File: Formula/xdg-ninja.rb (hunk 1 above)
Comment (superatomic) [reply Formula/xdg-ninja.rb:19]: Kind of. There's no machine readable information in the output, and output isn't guaranteed to stay the constant between updates.

However, there are some sections of the output that will probably stay constant. I could check for those, if you want.

File: Formula/xdg-ninja.rb (hunk 2 above)
Comment (superatomic) [reply Formula/xdg-ninja.rb:13]: I didn't really know where to put everything. The important part is that `"programs/" and `"xdg-ninja.sh"` must be installed in the same directory.

I can change `prefix` to `pkgshare` if that would be better.

Comment (carlocab) [reply Formula/xdg-ninja.rb:13]: Yea, `pkgshare` seems better.

**Comment by carlocab** [comment]
Well, we try to have formulae that are feature-complete which means catering to some niche use-cases (but not necessarily all of them).

If the build doesn't take too long and the `xdgnj` binary isn't too large, then it probably doesn't hurt to build it. That's assuming `xdgnj` is usable after cloning the `xdg-ninja` repository separately -- though I don't see why it wouldn't be. (It would be very silly for there to be restrictions about where the `xdgnj` binary is located on the filesystem relative to the `xdg-ninja` repository. After all, a developer could put it anywhere after building it from source upon cloning the repo.)

File: Formula/xdg-ninja.rb (hunk 1 above)
Comment (carlocab) [reply Formula/xdg-ninja.rb:19]: Hmmm, no, I guess this is fine.

File: Formula/xdg-ninja.rb (hunk 2 above)
Comment (superatomic) [reply Formula/xdg-ninja.rb]: Cool. I've pushed a commit with those changes.

**Comment by superatomic** [comment]
Great! I'll add the `xdgnj` binary to the formula, although it will be a few days before I will be able to implement it.

Should I also add the name "xdgnj" as a name alias for this formula?

**Comment by superatomic** [comment]
I tried to add `xdgnj` to the formula, but I run into the following error:

```
==> stack install --system-ghc --no-install-ghc --local-bin-path=/usr/local/Cellar/xdg-ninja/0.2.0.1/bin
Last 15 lines from /Users/Ethan/Library/Logs/Homebrew/xdg-ninja/01.stack:
2022-07-08 00:05:22 +0000

stack
install
--system-ghc
--no-install-ghc
--local-bin-path=/usr/local/Cellar/xdg-ninja/0.2.0.1/bin

No compiler found, expected minor version match with ghc-9.0.2 (x86_64) (based on resolver setting in /private/tmp/xdg-ninja-20220707-6745-h8ymcy/xdg-ninja-0.2.0.1/stack.yaml).
To install the correct GHC into /private/tmp/xdg-ninja-20220707-6745-h8ymcy/xdg-ninja-0.2.0.1/.brew_home/.stack/programs/x86_64-osx/, try running "stack setup" or use the "--install-ghc" flag. To use your system GHC installation, run "stack config set system-ghc --global true", or use the "--system-ghc" flag.
```

This is the additional line I have added to the `install` method in addition to adding `depends_on "ghc" => :build` and `depends_on "haskell-stack" => :build`:
```ruby
system "stack", "install", "--system-ghc", "--no-install-ghc", "--local-bin-path=#{bin}"
```

If I let `stack` install it's own version of `ghc` instead of using Homebrew's version (`ghc@9`) by removing `--system-ghc` and `--no-install-ghc` flags, the build will work, but then I'm pulling a resource from an external source to build, and I don't think that's allowed.

I don't actually know Haskell and I've never used `stack` (or `cabal`), so I don't know if there is an easy way to fix this, but I'd imagine that for this to work either `ghc-9.0.2` specifically has to be downloaded by the formula, or the dependency file needs to be patched to be less restrictive.

Unless somebody knows how to use a specific version of `ghc` or how to build successfully while using Homebrew's version of `ghc`, I'd recommend that `xdgnj` is simply not installed in this formula.

**Review by carlocab — APPROVED** [review]

**Review by chenrui333 — APPROVED** [review]

**Comment by chenrui333** [comment]
@superatomic, thanks for your contribution to Homebrew! :tada: 🥇

Without awesome contributors like you, it would be impossible to maintain Homebrew to the high level of quality users have come to expect. Thank you!!!!

**Pull request closed without merging**